### PR TITLE
Add ping option to FPS Control plugin

### LINKED ISSUE #2544: FPS counter not showing with stretched fixed mode on
title

### LINKED ISSUE #2640: Add ping indicator
An infobox like the Timers or Slayer counter that shows ping would be amazing

### PR DESCRIPTION
**Pull request by CustardGod**
Adds an indicator in the top right corner that shows the user's ping to the current world.

- Indicators automatically move out of the way of the log out button on resizeable mode
- Allows you to disable both FPS and Ping indicators

Text color is Green below 50, Yellow from 50 to 99, and Red above 100.
May be inaccurate at times until hook is added for finding host for the current world.

Fixes #2544 
Fixes #2640 

image

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsIndicatorLocation.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package net.runelite.client.plugins.fps;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsIndicatorLocation.java]: Why is this needed?

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsIndicatorLocation.java]: This is like purely debug information, i dont think we need fancy toggles

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java [hunk 2]
```diff
@@ -50,7 +51,9 @@
 	private static final int MAX_FPS = 50;
 	private static final int FPS_SIZE = MAX_FPS + 1;
 	private static final int Y_OFFSET = 14;
-	private static final int VALUE_X_OFFSET = 15;
+	private static int VALUE_X_OFFSET = 15;
+	private final PanelComponent panel = new PanelComponent();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java]: Move the panel component where it belongs, next to other instance field declarations

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java]: Also, this is unused?

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsPlugin.java [hunk 3]
```diff
@@ -46,13 +56,19 @@
  * game and system load, it usually finds the sweet spot in about two seconds.
  */
 @PluginDescriptor(
-	name = "FPS Control",
+	name = "Performance",
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsPlugin.java]: If you are renaming plugin, rename classes as well

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsIndicatorLocation.java (hunk 1 above)
Comment (LeviSchuck) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsIndicatorLocation.java]: It covers up the stretch mode X button thing in that mode

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsIndicatorLocation.java]: Well, then we can just detect that and move both overlays down

Comment (CustardGod) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsIndicatorLocation.java]: Yeah I added this because people complained about it covering the log out button in discord, not sure if an issue was ever posted though

Comment (CustardGod) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsIndicatorLocation.java]: I'm really not sure how I feel about that config option anyway. It blocks the left click option text so it's kind of out of place. I'll probably just update this to move a bit to the left when the logout button is in the corner

File: runelite-client/src/main/java/net/runelite/client/plugins/performance/PerformanceOverlay.java [hunk 4]
```diff
@@ -45,30 +46,35 @@
...
  * along with a locked position for the FPS value.
  */
-public class FpsOverlay extends Overlay
+public class PerformanceOverlay extends Overlay
 {
 	private static final int MAX_FPS = 50;
 	private static final int FPS_SIZE = MAX_FPS + 1;
 	private static final int Y_OFFSET = 14;
-	private static final int VALUE_X_OFFSET = 15;
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/performance/PerformanceOverlay.java]: No need for empty line

File: runelite-client/src/main/java/net/runelite/client/plugins/performance/PerformanceOverlay.java [hunk 5]
```diff
@@ -45,30 +46,35 @@
...
-	private final FpsConfig config;
+	private final PerformanceConfig config;
 	private final Client client;
 
+	@Inject
+	private PerformancePlugin plugin;
+
 	// Often changing values
 	private boolean isFocused = true;
+	private int VALUE_Y_AXIS_PING;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/performance/PerformanceOverlay.java]: Why are these 2 (VALUE_Y.., point) instance variables and not local in the render method?

Comment (CustardGod) [reply runelite-client/src/main/java/net/runelite/client/plugins/performance/PerformanceOverlay.java]: Changes made

**Comment by deathbeam** [comment]
This PR needs mostly general cleanup. I can maybe look at it tomorrow, or if anyone wants to clean it up a bit :d

File: runelite-client/src/main/java/net/runelite/client/plugins/performance/PerformancePlugin.java [hunk 6]
```diff
@@ -102,4 +121,47 @@ protected void shutDown() throws Exception
...
+				host = InetAddress.getByName("oldschool" + (client.getWorld() - 300) + ".runescape.com");
+			}
+			catch (UnknownHostException he)
+			{
+				log.warn("Cannot ping host", he);
+				return -1;
+			}
+
+			Instant start = Instant.now();
+			try (Socket sock = new Socket(host, 443))
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/performance/PerformancePlugin.java]: This would need a socket timeout.

**Comment by CustardGod** [comment]
By this you mean separate it into 2 PRs?

File: runelite-client/src/main/java/net/runelite/client/plugins/performance/PerformancePlugin.java (hunk 6 above)
Comment (CustardGod) [reply runelite-client/src/main/java/net/runelite/client/plugins/performance/PerformancePlugin.java]: Something like what I've just added?

**Comment by deathbeam** [comment]
Just 2 commits i guess

**Comment by Nightfirecat** [comment]
What Adam is asking is to have two commits that contain your full diff: one which does nothing except rename the plugin, and the other that does nothing but add the code for tracking and displaying ping. That way, there are no "clean-up" or "fixup" commits and the history will be clearer.

**Comment by CustardGod** [comment]
Would someone on discord mind explaining how to do this?

image

**Comment by CustardGod** [comment]
To save myself the headache of working around git which I have no idea how to use at this point I've changed the PR to only add Ping to the FPS Control plugin. If requested I'll change the name in a separate PR.

**Comment by LeviSchuck** [comment]
@CustardGod can you not include the merge branch commit?

**Comment by CustardGod** [comment]
Whenever I resolve this conflict it adds that commit

**Comment by CustardGod** [comment]
There we go, fixed it

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsConfig.java [hunk 7]
```diff
@@ -29,9 +29,9 @@
 import net.runelite.client.config.ConfigItem;
 
 @ConfigGroup(
-	keyName = FpsPlugin.CONFIG_GROUP_KEY,
-	name = "FPS Control",
-	description = "Lets you control what your game frame rate is, often helps keep CPU down too"
+		keyName = FpsPlugin.CONFIG_GROUP_KEY,
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsConfig.java]: What is with these extra spaces?

Comment (CustardGod) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsConfig.java]: A bunch of the config stuff was double indented like that and I tried to fix it, I guess I missed some. Mb

Comment (CustardGod) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsConfig.java]: Fixed

**Comment by CustardGod** [comment]
@Adam- @deathbeam Could I get you 2 to have a look over this again?

**Comment by LeviSchuck** [comment]
I've done some clean up, added a separate commit that renames the plugin, updated a bit of the javadocs.

**Comment by CustardGod** [comment]
Thanks for adding the separate commit for the name change, I could not for the life of me get it to work properly

**Comment by LeviSchuck** [comment]
This now passes automatic checks for merge.

I have done code cleanup on the primary commit which added ping to the FPS control plugin.

As requested, the plugin is being renamed to `Performance` Relevant classes have been renamed which are shared. FPS only classes have not been renamed. The package has been renamed.

**Comment by CustardGod** [comment]
Good stuff levi thanks for the help :D

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsPlugin.java [hunk 8]
```diff
@@ -102,4 +122,66 @@ protected void shutDown() throws Exception
 	{
 		drawManager.unregisterEveryFrameListener(drawListener);
 	}
+
+	@Schedule(
+		asynchronous = true,
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsPlugin.java]: This will block everything that is using the ScheduledExecutorService every 5 seconds, not sure if it should be done like this, maybe yes.

Comment (CustardGod) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsPlugin.java]: Do you have any suggestions to change it to? I was initially using 5 seconds just to make it so I didn't have to wait a long time for it to update while testing. I'm opening to changing it to a longer period if that's something you would like

Comment (AlexKolpa) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsPlugin.java]: Given that you're doing blocking network operations on it, I'd personally go with a separate single thread `ScheduledExecutorService` while draw ping is set. You'd need to do a slight bit of book-keeping for the starting and closing of the service, but it seems better over potentially blocking a lot of other operations using the main executor.

Comment (CustardGod) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsPlugin.java]: Would you mind taking a look at what I've added?

**Comment by CustardGod** [comment]
@Adam- @deathbeam Could I get you guys to look over this again? Once everything is fixed it'll be broken into the 2 commits you asked for before. I'd like to have this done for tomorrow

**Comment by LeviSchuck** [comment]
Rewrote history again so it's down to 2 logical commits.

**Comment by Adam-** [comment]
I am not sure how much Jagex would appreciate us doing all of these full tcp connections to them so frequently. I would rather do a real ping. Also this pr is timing the full 3way handshake and not the actual RTT. If we really wanted Ping, I would probably do something like this https://timothybasanov.com/2007/06/25/java-windows-ping.html (JNA+IcmpSendEcho) and then only support Windows.

**Comment by CustardGod** [comment]
Just looking through this, you said JNA+IcmpSendEcho but isn't that JNI not JNA? I haven't used either so I don't know the difference

**Comment by CustardGod** [comment]
While I'm thinking of it, what would you think of just using Runtime.exec() to ping and just using regex to get the result instead? Seems like a simple way of getting it to work on other operating systems

**Comment by CustardGod** [comment]
Updated to not make a full tcp connection. Instead uses Runtime.exec() ping command depending on the user's OS. Keeps this plugin from only working on Windows.

**Comment by CustardGod** [comment]
It seems odd to me that using JNI for this is considered cleaner than what we have currently. It seems like an over the top solution to a problem that we don't have anymore and causes problems itself being that it will only work on Windows. I have no idea how to properly use JNI and I'd rather not have to spend the time learning how to use it plus learn a C-compatible language. I have to go for the night @Adam- so I don't have any more time to look into this for now. Please explain this to me past the  "I think it's the cleanest solution" response you gave me on discord so I can better understand the situation. Thanks.
